### PR TITLE
Adds `:onhoverlost` and separates out event based actions on eventbox widget (See #251 for more info).

### LINKED ISSUE #250: [FEATURE] Hover out events?
## Description of the requested feature
Every widget has `:onhover` but not `:onhoverout` (or something like that, don't mind the name). [Gtk](https://people.gnome.org/~shaunm/girdoc/C/Gtk.leave-notify-event.html) issues leave notify event, so should be supported in eww as well.

In most of cases like hovering to open the revealer, we also want to close it while leaving the object (or its children).

## Proposed configuration syntax
`(somewidget :onhoverout "cmd to run when hover finishes" ...)`

# Additional context
\-

### LINKED ISSUE #251: [BUG] Event based actions doesn't work on most widgets
## Checklist before submitting an issue
- [x] I have searched through the existing [closed and open issues](https://github.com/elkowar/eww/issues?q=is%3Aissue) for eww and made sure this is not a duplicate
- [x] I have specifically verified that this bug is not a common [user error](https://github.com/elkowar/eww/issues?q=is%3Aissue+label%3Ano-actual-bug+is%3Aclosed)
- [x] I am providing as much relevant information as I am able to in this bug report (Minimal config to reproduce the issue for example, if applicable)

## Describe the bug
Most of listening properties don't work, like `:onhover`, `:cursor`, etc. due to wrong implementation (listening to wrong events).

Listen to leave notify event without leave event mask ([EventMask::LEAVE_NOTIFY_MASK](https://gtk-rs.org/gtk3-rs/stable/latest/docs/gdk/struct.EventMask.html#associatedconstant.LEAVE_NOTIFY_MASK)).
https://github.com/elkowar/eww/blob/70285e0ebf928df4d78725dfb02d9eae2273630c/crates/eww/src/widgets/widget_definitions.rs#L157

Listen to button press event without masking to button press event ([EventMask::BUTTON_PRESS_EVENT](https://gtk-rs.org/gtk3-rs/stable/latest/docs/gdk/struct.EventMask.html#associatedconstant.BUTTON_PRESS_MASK)). Although this one works somehow 🤔
https://github.com/elkowar/eww/blob/70285e0ebf928df4d78725dfb02d9eae2273630c/crates/eww/src/widgets/widget_definitions.rs#L447-L449

Listen to value change event without masking property change mask ([EventMask::PROPERTY_CHANGE_MASK](https://gtk-rs.org/gtk3-rs/stable/latest/docs/gdk/struct.EventMask.html#associatedconstant.PROPERTY_CHANGE_MASK))
https://github.com/elkowar/eww/blob/70285e0ebf928df4d78725dfb02d9eae2273630c/crates/eww/src/widgets/widget_definitions.rs#L205-L207

## Reproducing the issue

```
(box :class "cyan" :cursor "text" :onhover "kitty" "some text to make box visible")
```
Try hovering, neither cursor changes, nor command runs.

## Expected behaviour
Self explainatory, should do what is expected.

## Additional context

https://user-images.githubusercontent.com/26714676/131550687-6b8d3828-2d92-4212-ab06-1d2385ad136e.mp4

### PR DESCRIPTION
**Pull request by Animeshz**
## Description

Deprecates `:timeout`, `:onscroll`, `:onhover` and `:cursor` from top of widget hierarchy.
Moves these to a new widget `(eventbox ...)`.
Adds `:onhoverlost` property to `(eventbox ...)`.

Fix #250 and fix #251 (see them for more details).

## Usage

```
(eventbox :timeout <timeout>
          :onscroll "<cmd to run on scroll with {} being replaced by up or down>"
          :onhover "<cmd to run on hover with {} being replaced by 'x y' coordinate>"
          :onhoverlost "<cmd to run on hover loss wih {} being replaced by 'x y' coordinate>"
          :cursor "<type of cursor>"
    "Content")
```

### Showcase

https://user-images.githubusercontent.com/26714676/131639040-89781f4e-68af-4079-be2b-0c6ac4ff6d97.mp4

Edit: I should have probably add revealer within the box contained in the eventbox widget, so that hover applies to both revealer and the date here to remain it opened. Ignore it please 😅😅

## Additional Notes

Widget must contain exactly one child (restriction from GTK).

## Checklist

Please make sure you can check all the boxes that apply to this PR.

- [x] All widgets I've added are correctly documented.
- [x] The documentation in the `docs/content/main` directory has been adjusted to reflect my changes.
- [x] I used `cargo fmt` to automatically format all code before committing

### CODE REVIEW

**Comment by Animeshz** [comment]
https://user-images.githubusercontent.com/26714676/131776584-e7c5794c-6051-4686-85b5-b1a8c2b3abd5.mp4

There seem to be a glitch when moving into the revealer 🤔🤔, even if its part of it, it tries to hide for some reasons. What it could be?

```
(defwidget barmiddle []
  (eventbox :cursor "text" :onhover "eww update playerctl=\"true\"" :onhoverlost "eww update playerctl=\"false\""
    (box :orientation "h" :space-evenly false :halign "center" :spacing 0
      (revealer :transition "slideleft" :duration "1s" :reveal playerctl
        (box :orientation "h" :space-evenly false :spacing 0
          (music) (separator)))
      (date_time))))
```

Any idea?

**Comment by Animeshz** [comment]
Glitch is more clear when we have the revealer open with this cursor change

https://user-images.githubusercontent.com/26714676/131777281-d425b5b4-3552-4dd0-9dfa-d66620e71679.mp4

**Comment by Animeshz** [comment]
Applied fix (donot discriminate nested elements for hover events):

https://user-images.githubusercontent.com/26714676/131795837-d9ef5d3b-a8ab-46cf-aefe-e842e47e6316.mp4

I guess PR is ready now, ping me if there's any problem remaining!

**Comment by snakedye** [comment]
I would replace "hoverlost" by leave also moving those to a different widget is a big change. Is there any reason why there actually need to be a new widget?

**Comment by Animeshz** [comment]
@snakedye See #251 GTK does not sent events to most of the widgets, its inconsistent behavior.

**Comment by elkowar** [comment]
If using eventbox is genuinely the only way to support these events across all widgets, then I fear this _is_ the best way to go about it. auto-wrapping widgets that use any of these properties would make debugging more confusing and definitely be complex to implement,...

The deprecation explanation definitely should give a bit more details as to what eventbox does, _why_ it's necessary, and optimally link to the related issue. An explanation such as 
> Using the attribute `:cursor` here has been removed, as GTK does not support it consistently. Instead, use `eventbox` to wrap this widget and set the attribute there. See https://github.com/elkowar/eww/issues/251 for more details

would make this a lot cleaner.

**Comment by Animeshz** [comment]
> auto-wrapping widgets that use any of these properties would make debugging more confusing and definitely be complex to implement

I thought of auto-wrapping, but changed decision to just let user manually wrap for the same reason of easier debugging from end user, also it looks clean and easier to know which positions events are going to be captured in debugger as well as config file via this...

I also couldn't figure out a better way to establish a consistent behavior among all widgets which changes their support with runtime versions of gtk...

> would make this a lot cleaner.

👍

---

Also what do you prefer (is better option to have) `:onhoverlost` or `:onleave`?

**Comment by elkowar** [comment]
given that the name `onhover` is better matched by a `onhoverlost`, i'd prefer that

File: crates/eww/src/widgets/widget_definitions.rs [hunk 1]
```diff
@@ -47,10 +56,25 @@ pub(super) fn widget_to_gtk_widget(bargs: &mut BuilderArgs) -> Result<gtk::Widge
...
 /// @widget widget
 /// @desc these properties apply to _all_ widgets, and can be used anywhere!
 pub(super) fn resolve_widget_attrs(bargs: &mut BuilderArgs, gtk_widget: &gtk::Widget) {
+    let deprecated: HashSet<_> = DEPRECATED_ATTRS.to_owned();
+    let contained_deprecated: Vec<_> = bargs.unhandled_attrs.drain_filter(|a| deprecated.contains(&a.0 as &str)).collect();
+    if !contained_deprecated.is_empty() {
+        let diag = error_handling_ctx::stringify_diagnostic(gen_diagnostic! {
+            kind =  Severity::Error,
+            msg = format!("The attribute(s) ({}) has/have been removed, as GTK does not support it consistently. Instead, use eventbox to wrap this widget and set the attribute there. See #251 (https://github.com/elkowar/eww/issues/251) for more details.", contained_deprecated.iter().join(", ")),
+            label = bargs.widget.span => "Found in here"
```
Comment (elkowar) [inline crates/eww/src/widgets/widget_definitions.rs:73]: The long text should probably better be in a note, allowing the `msg` itself to be shorter (i.e. just `Unsupported properties provided`). The gen_diagnostic! macro has a note field for that

**Comment by elkowar** [comment]
Otherwise, some conflicts, but otherwise this looks good!

**Comment by Animeshz** [comment]
For some reason gtk 0.14 update was creating a lot of conflict in other files as well, gimme some time to resolve them 😅😅

**Comment by elkowar** [comment]
looks good now! thanks a lot!

**Pull request merged**